### DISCUSSION TITLE
model_validator(mode='after') in combination with mypy

### DISCUSSION THREAD

**Discussion by casoku** [pydantic/pydantic#6502] [Pydantic V2]
Hello all,

Recently I tried to upgrade from pydantic V1 to V2 (pydantic-2.0.2), but when I replace root_validators(pre='false') with model_validator(mode='after') mypy fails. This also seem to happen when I use the example code.

I use plain (no configuration) mypy V1.4.1 and python 3.10.10 to check the following code:

```python
from pydantic import BaseModel, ValidationError, model_validator


class UserModel(BaseModel):
    username: str
    password1: str
    password2: str

    @model_validator(mode='before')
    def check_card_number_omitted(cls, data):
        assert 'card_number' not in data, 'card_number should not be included'
        return data

    @model_validator(mode='after')
    def check_passwords_match(cls, m: 'UserModel'):
        pw1 = m.password1
        pw2 = m.password2
        if pw1 is not None and pw2 is not None and pw1 != pw2:
            raise ValueError('passwords do not match')
        return m
```

I get this error:
```python
error: Argument 1 has incompatible type "Callable[[UserModel, UserModel], UserModel]"; expected "ModelAfterValidator | ModelAfterValidatorWithoutInfo"  [arg-type]
```

Using the mode='before', does not give this issue since it seem to have less strict typing. Is it possible to use mode='after' in combination with mypy? Or do I need to do something different?

**Comment by hramezani** [comment]
Hello @casoku,

I think this is a duplicate of https://github.com/pydantic/pydantic/issues/6277

There is a [mypy issue for this](https://github.com/python/mypy/issues/15620)

**Comment by casoku** [comment]
Ah, I must have missed it, my mistake. Thanks for your response.
